src/builtins.c

```diff
@@ -659,29 +660,71 @@ DLLEXPORT int jl_substrtod(char *str, size_t offset, int len, double *out)
         bstr = newstr;
         pend = bstr+len;
     }
-    *out = strtod_c(bstr, &p);
-    if ((p == bstr) || (p != pend) ||
-        (errno==ERANGE && (*out==0 || *out==HUGE_VAL || *out==-HUGE_VAL)))
-        err = 1;
+    double out = strtod_c(bstr, &p);
+    err = errno;
+
     if (bstr != str+offset)
         free(bstr);
-    return err;
+
+    if (err==ERANGE && (out==0 || out==HUGE_VAL || out==-HUGE_VAL)) {
+        errno = ERANGE;
+        err = 1;
+    }
+    else if ((p == bstr) || (p != pend)) {
+        err = 3;
+    }
+    else {
+        err = 0;
+    }
+
+    return (jl_nullable_float64_t){(uint8_t)err, out};
 }
 
-DLLEXPORT int jl_strtod(char *str, double *out)
+DLLEXPORT jl_nullable_float64_t jl_maybe_strtod(char *str)
 {
     char *p;
+    int err = 0;
+
     errno = 0;
-    *out = strtod_c(str, &p);
-    if (p == str ||
-        (errno==ERANGE && (*out==0 || *out==HUGE_VAL || *out==-HUGE_VAL)))
-        return 1;
-    while (*p != '\0') {
-        if (!isspace((unsigned char)*p))
-            return 1;
-        p++;
+    double out = strtod_c(str, &p);
+
+    if (errno==ERANGE && (out==0 || out==HUGE_VAL || out==-HUGE_VAL)) {
+        err = 1;
+    }
+    else if (p == str) {
+        err = 3;
+    }
+    else {
+        while (*p != '\0') {
+            if (!isspace((unsigned char)*p)) {
+                err = 3;
```

Comment:
I don't think we should put an integer error code in a Bool field. Asking for trouble.